core/business.go

```diff
@@ -172,29 +311,89 @@ func NewDispatcher(maxWorkers, maxQueue int, mfile string, minterval int64) *Dis
 	AgentMetrics = Metrics{In: inT, Failed: failT, Total: totT, TotalBytes: totB, Bytes: bytesT}
 	go metrics.Log(r, time.Duration(minterval)*time.Second, log.New(f, "metrics: ", log.Lmicroseconds))
 
-	// define pool of workers and jobqueue
-	pool := make(chan chan Job, maxWorkers)
-	JobQueue = make(chan Job, maxQueue)
-	return &Dispatcher{JobPool: pool, MaxWorkers: maxWorkers}
+	if TransferType == "pull" {
+		StorageQueue = make(chan Job, storageQueueSize)
+		RequestQueue = make(PriorityQueue, 0) // Create a priority queue
+		// Load pending requests from DB
+		heap.Init(&RequestQueue)
+		requests, err := TFC.ListRequest("pending") // Load requests from database
+		check("Unable To fetch data", err)
+		for i := 0; i < len(requests); i++ {
+			heap.Push(&RequestQueue, &Item{Value: requests[i], priority: requests[i].Priority})
+		}
+		logs.Println("Requests restored from db")
+	}
+
+	TransferQueue = make(chan Job, transferQueueSize)
+}
+
+// Run function starts the worker and dispatch it as go-routine
+func (d *Dispatcher) StorageRunner() {
+	// starting n number of workers
+	for i := 0; i < d.MaxWorkers; i++ {
+		worker := NewWorker(i, d.JobPool)
+		worker.Start()
+	}
+
+	go d.dispatchToStorage()
+}
+
+func (d *Dispatcher) dispatchToStorage() {
+	for {
+		select {
+		case job := <-StorageQueue:
+			// a job request has been received
+			go func(job Job) {
+				// try to obtain a worker job channel that is available.
+				// this will block until a worker is idle
+				jobChannel := <-d.JobPool
+
+				// dispatch the job to the worker job channel
+				jobChannel <- job
+			}(job)
+		default:
+			time.Sleep(time.Duration(10) * time.Millisecond) // wait for a job
+		}
+	}
 }
 
 // Run function starts the worker and dispatch it as go-routine
-func (d *Dispatcher) Run() {
+func (d *Dispatcher) TransferRunner() {
 	// starting n number of workers
 	for i := 0; i < d.MaxWorkers; i++ {
 		worker := NewWorker(i, d.JobPool)
 		worker.Start()
 	}
 
-	go d.dispatch()
+	go d.dispatchToTransfer()
 }
 
-func (d *Dispatcher) dispatch() {
+func (d *Dispatcher) dispatchToTransfer() {
 	for {
 		select {
-		case job := <-JobQueue:
+		case job := <-TransferQueue:
 			// a job request has been received
 			go func(job Job) {
+				if TransferType == "pull" {
+					// Update the status of request in DB
+					err, status := TFC.GetStatus(job.TransferRequest.Id)
+					if status == "pending" {
+						err = TFC.UpdateRequest(job.TransferRequest.Id, "processing")
+					} else {
+						return
+					}
+					err = TFC.RetriveRequest(&job.TransferRequest)
+					if err != nil {
+						// TODO: push in error queue.
+						return
```

Comment:
which status transfer request will have here when you return, should you first update the job.TransferRequest.Status = "error"? The same argument applies to line 395